core/src/test/scala/neotypes/FutureAsyncSpec.scala

```diff
@@ -0,0 +1,92 @@
+package neotypes
+
+import org.scalatest.{FlatSpec, Matchers}
+
+import scala.concurrent.{ExecutionContext, Future}
+import org.scalatest.concurrent.ScalaFutures
+class FutureAsyncSpec extends FlatSpec with Matchers {
+  implicit val ec = ExecutionContext.Implicits.global
+
+  trait FutureAsyncFixture {
+    var counter = 0
+    var resourceIsOpen = true
+    val fa = Future.successful(1)
+    val f: Int => Future[String] = i => Future.successful(i.toString)
+    val finalizer = { (_: Int, _: Option[Throwable]) => {Future.successful {resourceIsOpen = false}}}
+  }
+
+  "async future" should "execute finalizer" in new FutureAsyncFixture{
+    ScalaFutures.whenReady(Async.futureAsync.guarantee(fa)(f)(finalizer)){
+      w =>
+        w shouldBe "1"
+        resourceIsOpen shouldBe false //resource is closed
+    }
+  }
+
+  "async future" should "execute finalizer when fa fails and return the exception" in new FutureAsyncFixture{
+    override val fa = Future.failed(new Exception("Boom!"))
+
+    ScalaFutures.whenReady(Async.futureAsync.guarantee(fa)(f)(finalizer).failed){
+      ex =>
+        ex.getMessage shouldBe "Boom!"
+        resourceIsOpen shouldBe true //resource is closed
+    }
+  }
+
+  "async future" should "execute finalizer when f fails and return the exception" in new FutureAsyncFixture{
+    val runtimeEx = new Exception("Boom!")
+    override val f: Int => Future[String] = _ => Future.failed(runtimeEx)
+
+    ScalaFutures.whenReady(Async.futureAsync.guarantee(fa)(f)(finalizer).failed){
+      ex =>
+        ex shouldBe runtimeEx
+        resourceIsOpen shouldBe false //resource was closed
+    }
+  }
+
+  "async future" should "execute finalizer and return f exception when f and finalizer fail" in new FutureAsyncFixture {
+    val runtimeEx = new Exception("Boom!")
+    override val f: Int => Future[String] = _ => Future.failed(runtimeEx)
+    override val finalizer = { (_: Int, _: Option[Throwable]) => {Future.failed(new Exception("Finalizer Failed!"))} }
+
+    ScalaFutures.whenReady(Async.futureAsync.guarantee(fa)(f)(finalizer).failed){
+      ex =>
+        ex.getMessage shouldBe "Boom!"
+        resourceIsOpen shouldBe true //resource was not closed
+    }
+  }
+
+  "async future" should "return finalizer exception" in new FutureAsyncFixture {
+    override val finalizer = { (_: Int, _: Option[Throwable]) => {
+      counter = counter + 1
+      Future.failed(new Exception(s"Finalizer failure count: ${counter}!"))
+      }
+    }
+
+    ScalaFutures.whenReady(Async.futureAsync.guarantee(fa)(f)(finalizer).failed){
+      ex =>
+        counter shouldBe 2
```

Comment:
I belive `counter` should be `1`.

Reply:
I think finalizer is called once here:
```java
_ <- finalizer(a, None)
```

and then again here:

```java
case ex: Throwable =>
              finalizer(a, Some(ex))
```

Reply:
Yeah but it shouldn't. That is the point.

Reply:
If you revert the code change and run this test with original code, I think you'll find the behavior is the same

Reply:
Could you check that? In any case, we need to revert the changes since the actual implementation is buggy, you can see that other tests like the refined one fail.

I believe we could simplify the previous implementation using a `transformWith`, those tests would ensure that whatever change we made are secure, which is fantastic!

Reply:
I wrote the test against the original functionality, and then made my code change, so I am very sure I didn't break anything.  I will revert it for the PR, however.

Reply:
reverted the code change to Async